### PR TITLE
Support of `RealJenkinsRule` in `InboundAgentRule`

### PR DESCRIPTION
**Pull request by Vlatombe**
Add several methods to :
* use this rule in the context of `RealJenkinsRule`
* use this rule in the context of CloudBees way of starting up agents.

See downstream usage in https://github.com/jenkinsci/jenkins/pull/7381

- [X] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [X] Ensure that the pull request title represents the desired changelog entry
- [X] Please describe what you did
- [ ] Link to relevant issues in GitHub or Jira
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [ ] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

File: src/main/java/org/jvnet/hudson/test/InboundAgentRule.java [hunk 1]
```diff
@@ -249,44 +269,40 @@ public void start(@NonNull JenkinsRule r, Options options) throws Exception {
...
         if (options.isSecret()) {
-            String secret = c.getJnlpMac();
             // To watch it fail: secret = secret.replace('1', '2');
-            cmd.addAll(Arrays.asList("-secret", secret));
+            cmd.addAll(Arrays.asList("-secret", agentArguments.secret));
         }
-        JNLPLauncher launcher = (JNLPLauncher) c.getLauncher();
-        if (!launcher.getWorkDirSettings().isDisabled()) {
-            cmd.addAll(launcher.getWorkDirSettings().toCommandLineArgs(c));
+        if (agentArguments.commandLineArgs != null) {
```
Comment (jglick) [inline src/main/java/org/jvnet/hudson/test/InboundAgentRule.java:281]: Could make this non-null.

**Review by jglick — APPROVED** [review]

**Review by basil — COMMENTED** [review]
LGTM. Any downstream PRs need filing to adapt consumers in Jenkins core?

**Comment by basil** [comment]
Just because it is compatible does not mean that existing callers do not need updating.

**Comment by jglick** [comment]
True, it looks like `JNLPLauncherRealTest` could be simplified based on this.

**Comment by Vlatombe** [comment]
I missed `JNLPLauncherRealTest` when checking for downstream usage, I'll update it.

File: src/main/java/org/jvnet/hudson/test/InboundAgentRule.java [hunk 2]
```diff
@@ -321,4 +337,134 @@ public boolean isAlive(String name) {
...
+            }
+        }
+    }
+
+    private static class GetNumberOfNodes implements RealJenkinsRule.Step2<Integer> {
+        @Override
+        public Integer run(JenkinsRule r) throws Throwable {
+            return r.jenkins.getNodes().size();
+        }
+    }
```
Comment (basil) [inline src/main/java/org/jvnet/hudson/test/InboundAgentRule.java:436]: What is the intended usage pattern for this step? I cannot see any usages in https://github.com/jenkinsci/jenkins/pull/7381.

Comment (Vlatombe) [reply src/main/java/org/jvnet/hudson/test/InboundAgentRule.java]: Indeed, I forgot to remove it after refactoring. done.

File: src/main/java/org/jvnet/hudson/test/InboundAgentRule.java [hunk 3]
```diff
@@ -321,4 +337,127 @@ public boolean isAlive(String name) {
...
+        @Override
+        public void run(JenkinsRule r) throws Throwable {
+            stop(r, name);
+        }
+
+        private static void stop(JenkinsRule r, String name) throws InterruptedException {
+            Computer c = r.jenkins.getComputer(name);
+            if (c != null) {
+                while (c.isOnline()) {
+                    Thread.sleep(100);
```
Comment (jglick) [inline src/main/java/org/jvnet/hudson/test/InboundAgentRule.java:425]: The naming implies that this stops the agent, but really it just waits for it go offline?

**Review by jglick — APPROVED** [review]
Had a question on the downstream PR, but otherwise this looks ready to release to me.

File: src/main/java/org/jvnet/hudson/test/InboundAgentRule.java [hunk 4]
```diff
@@ -321,4 +337,127 @@ public boolean isAlive(String name) {
...
+        @Override
+        public String run(JenkinsRule r) throws Throwable {
+            createAgent(r, options);
+            return options.getName();
+        }
+
+        @SuppressFBWarnings(value = {"PATH_TRAVERSAL_IN", "NP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE"}, justification = "just for test code")
+        private static Slave createAgent(JenkinsRule r, Options options) throws Descriptor.FormException, IOException, InterruptedException {
+            if (options.getName() == null) {
+                Integer numberOfNodes = r.jenkins.getNodes().size();
```
Comment (basil) [inline src/main/java/org/jvnet/hudson/test/InboundAgentRule.java:447]: This wrapper type could be a primitive:

```suggestion
                int numberOfNodes = r.jenkins.getNodes().size();
```

Per IntelliJ:

> in some cases, boxing can be source of significant performance penalty, especially in loops.

**Review by basil — APPROVED** [review]
Thanks for the PR!

File: src/main/java/org/jvnet/hudson/test/InboundAgentRule.java [hunk 5]
```diff
@@ -321,4 +337,126 @@ public boolean isAlive(String name) {
...
+                options.setName("agent" + r.jenkins.getNodes().size());
+            }
+            JNLPLauncher launcher = new JNLPLauncher(true);
+            launcher.setWebSocket(options.isWebSocket());
+            DumbSlave s = new DumbSlave(options.getName(), new File(r.jenkins.getRootDir(), "agent-work-dirs/" + options.getName()).getAbsolutePath(), launcher);
+            s.setRetentionStrategy(RetentionStrategy.NOOP);
+            r.jenkins.addNode(s);
+            // SlaveComputer#_connect runs asynchronously. Wait for it to finish for a more deterministic test.
+            Computer computer = s.toComputer();
+            while (computer == null || computer.getOfflineCause() == null) {
```
Comment (jglick) [inline src/main/java/org/jvnet/hudson/test/InboundAgentRule.java:456]: https://github.com/jenkinsci/jenkins-test-harness/pull/558#discussion_r1120227068

**Pull request merged**